### PR TITLE
[IO-596] Add DeleteFiles utility class

### PR DESCRIPTION
**Pull request by jvz**
This class provides configurable strategies to delete files and directories.

Adapted from the work done in https://github.com/jenkinsci/jenkins/pull/3812.

Signed-off-by: Matt Sicker <boards@gmail.com>

@reviewbybees @garydgregory @jeffret-b

### CODE REVIEW

File: src/main/java/org/apache/commons/io/DeleteFiles.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+            final List<IOException> errors = tryDeleteRecursive(path);
+            if (errors.isEmpty()) {
+                return;
+            }
+            accumulatedErrors.addAll(errors);
+            if (waitToRetry(retriesAttempted)) {
+                throw new CompositeIOException(failureMessage(path, retriesAttempted, true), accumulatedErrors);
+            }
+        }
+        throw new CompositeIOException(failureMessage(path, maxRetries, false), accumulatedErrors);
```
Comment (asciborek) [inline src/main/java/org/apache/commons/io/DeleteFiles.java]: From what I can see this code will always cause throwing a new exception

Comment (jvz) [reply src/main/java/org/apache/commons/io/DeleteFiles.java]: Line 80 with the `return` will terminate the loop. This only returns successfully if none of the subpaths threw an exception.

File: src/main/java/org/apache/commons/io/CompositeIOException.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    public CompositeIOException(final String message, final IOException... causes) {
+        super(message);
+        this.causes = causes.clone();
+    }
+
+    public IOException[] getCauses() {
+        return causes.clone();
+    }
+}
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/CompositeIOException.java:44]: Note that we have `IOExceptionList`.

**Comment by garydgregory** [comment]
Some of this PR is likely obsolete due to the newer PathUtils delete methods. Please review these and perhaps propose enhancements?

**Comment by garydgregory** [comment]
Closing this PR, use version 2.7's `PathUtils` methods:
- `delete(Path)`
- `delete(Path, DeleteOption...)`
- `deleteDirectory(Path)`
- `deleteDirectory(Path, DeleteOption...)`
- `deleteFile(Path)`
- `deleteFile(Path, DeleteOption...)`

**Comment by jvz** [comment]
Breaking up into execution strategies sounds like a good idea! Such a strategy could likely adopt a similar API to retry libraries like retry4j.

**Comment by garydgregory** [comment]
@jvz 
Nice idea Matt :-) let's discuss on the ML where a retry API belongs.

**Comment by garydgregory** [comment]
On Wed, Sep 2, 2020 at 10:57 AM Matt Sicker <notifications@github.com>
wrote:

> Breaking up into execution strategies sounds like a good idea! Such a
> strategy could likely adopt a similar API to retry libraries like retry4j.
>

I am cutting 2.8.0-RC2 now, so let's think about this for 2.9.0.

Gary

**Comment by garydgregory** [comment]
Another item: values like `waitBetweenRetriesMillis` are error prone in my experience. Over at HttpComponents, we have classes called TimeValue and Timeout that wrap a `long` and a `TimeUnit`, and I'd like to bring that in to Apache Commons, someplace...

**Comment by garydgregory** [comment]
Some of this is superseded by the new(ish) code in PathUtils.delete methods and associated utilities. If we want a retry (count/duration delay) feature, let's build it on top of the newer code.

**Pull request closed without merging**